lib/elastic_apm/normalizers/rails/active_record.rb

```diff
@@ -19,26 +20,27 @@ def initialize(*args)
         end
 
         def normalize(_transaction, _name, payload)
-          return :skip if %w[SCHEMA CACHE].include?(payload[:name])
+          return :skip if SKIP_NAMES.include?(payload[:name])
 
           name = summarize(payload[:sql]) || payload[:name]
           context =
             Span::Context.new(db: { statement: payload[:sql], type: 'sql' })
-          [name, TYPE, subtype, ACTION, context]
+          [name, TYPE, subtype(payload), ACTION, context]
         end
 
         private
 
-        def subtype
-          @subtype ||= (lookup_adapter || 'unknown')
+        def subtype(payload)
+          @subtype ||= (lookup_adapter(payload) || 'unknown')
```

Comment:
This is still using `||=`. Is that what we want? If so, then is there in the end any reason to get the connection from the payload?

Reply:
I realise the overhead of checking, but if we're still just checking once, is there really a need? Do we need to memoize scoped on some key from the payload/connection?

Reply:
Good point, I missed that.
I think in Rails 6, the connection object could be different each time this method is reached, so you would want to check the payload object.
What about this? (**untested**)

```
def subtype(payload)
  return @subtype if @subtype
  if payload[:connection]
    lookup_adapter(payload)
  else
    @subtype = lookup_adapter(payload)
  end
end
```

That way, you'll continue looking up the adapter in the `payload` while it is present, otherwise, it'll memoize the `@subtype`.

Reply:
That's not totally true if `payload[:connection]` is only sometimes present?

```ruby
def subtype(payload)
  return lookup_adapter(payload[:connection]) if payload.key?(:connection)

  @default_adapter ||= lookup_adapter(::ActiveRecord::Base.connection)
end
```

`lookup_adapter` is the wrong method name in this version. Perhaps `adapter_from_connection`.

Reply:
Sorry, I should have explained that this is kind of a compromise - we will always looks up the `payload[:connection]` value if it exists but the first time it doesn't exist, we are going to memoize the result and not check anymore. I doubt Rails will sometimes have the `connection` in the payload and sometimes not (I'll check this), therefore I think it will never get to memoize the value. So this allows us to have different lookup methods for Rails 6 and < Rails 6 without us checking the Rails version.

Reply:
I just went through the active_support notifications source code and I can't find where the `connection` key is set so I don't know if it's always there for a `sql.active_record` event. I'll keep looking but maybe it makes more sense to do what you suggested.

Reply:
Update: I found where the `connection` is set in the payload. It's done [here](https://github.com/rails/rails/blob/966d3a7bf2c08c5c20213d29686cbb409a07734a/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L682) in the`AbstractAdapter`. So we can count on it always being there in Rails 6.0 for `sql.active_record` events.

Reply:
Wasn't one purpose of this to allow folks to have multiple adapters?

Reply:
The adapter will only ever be memoized if the payload doesn't have a `connection` key. Which apparently will never be the case with Rails 6.
So yes, people can have multiple adapters because the `connection` value in the payload will always be checked.

Reply:
But in any case, if the code isn't showing that without explanation, we should make it clearer = )
I'll incorporate your suggestion and change the method name to be more accurate.

Reply:
You're right and thanks. We could still memoize the `adapter => formatted string` method scoped on some (cheap to look up) key on `connection`.